### ISSUE TITLE
`misc-const-correctnes.WarnPointersAsValues` not reported for all variables in multi-declaration

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#57933]
```cpp
class A;

extern const A* cb();

void f()
{
    const A *a1 = cb(), *a2 = cb(); // no a2 warning

    const A * const a3 = cb(), *a4 = cb();
}
```

```
<source>:7:5: warning: variable 'a1' of type 'const A *' can be declared 'const' [misc-const-correctness]
    const A *a1 = cb(), *a2 = cb(); // no a2 warning
    ^
<source>:9:5: warning: variable 'a4' of type 'const A *' can be declared 'const' [misc-const-correctness]
    const A * const a3 = cb(), *a4 = cb();
    ^
```
https://godbolt.org/z/aK8o1Gfnh

I guess it's a known limitation but it doesn't appear to be documented,

CC'ing @JonasToth

**Comment by JonasToth** [comment]
that is interesting.
When marking the pointers `const` it works. It might be the case, that the warnings are generated, but deduplicated by the diagnostics system.
As they come from the same check and point to the same location, one of the warnings might go missing.

the limitation is only about fixing the issue. The warning itself should still be generated for both variables. But `fixit` is only emitted for single-declarations.

**Comment by firewave** [comment]
The warning for `a4` is no longer reported. This regressed with clang-tidy-21. Most likely caused by #130494. @HerrCai0907

> It might be the case, that the warnings are generated, but deduplicated by the diagnostics system.

@JonasToth Is it possible to disable the de-duplication? I recently added such a flag to Cppcheck and helped to uncovered some (unknown) issues.

**Comment by JonasToth** [comment]
@firewave There should be two occasions of deduplication (it has been a long time, please take this with grains of salt!)

`clang-tidy` does not emit the same diagnostic multiple times. This should be done by clangs diagnostic framework. E.g. templated code may result in the same warning over and over again - if the type is not part of the message - and they are not emitted by clang-tidy.

The second location is in `clang-apply-replacements`. I checked the code now and the function `mergeAndDeduplicate` deduplicates identical diagnostics emitted from different TUs (~ diagnostics in header files that are included in multiple .cpp files).
The code-path in `clang-apply-replacements` seems to be hard-coded and not configurable. But that is the best part to start, I guess.

**Comment by firewave** [comment]
@JonasToth Thanks for the fast and thorough reply.

> I checked the code now and the function `mergeAndDeduplicate` deduplicates identical diagnostics emitted from different TUs (~ diagnostics in header files that are included in multiple .cpp files).

That's what Cppcheck does as well but the way it determines if a finding is unique is flawed (not all data is considered - i.e. column and error ID) and leads to unique warnings to be suppressed (it is actually also somewhat abused). It seems like Clang might have the same shortcoming (it might be intentional though).

It seems like a minor thing since it is resolved after a roadtrip but depending on the code it might not just be a single one and might become quite an annoyance - especially if you if it requires you to go through something like a CI.
